lib/Dialect/FIRRTL/Transforms/GrandCentralTaps.cpp

```diff
@@ -0,0 +1,627 @@
+  Value value;
+  Annotation anno;
+};
+
+/// A data structure that caches and provides absolute paths to module instances
+/// in the IR.
+struct InstancePaths {
+  /// The instance graph of the IR.
+  InstanceGraph &instanceGraph;
+
+  explicit InstancePaths(InstanceGraph &instanceGraph)
+      : instanceGraph(instanceGraph) {}
+  ArrayRef<InstancePath> getAbsolutePaths(Operation *op);
+
+private:
+  /// An allocator for individual instance paths and entire path lists.
+  llvm::BumpPtrAllocator allocator;
+
+  /// Cached absolute instance paths.
+  DenseMap<Operation *, ArrayRef<InstancePath>> absolutePathsCache;
+
+  /// Append an instance to a path.
+  InstancePath appendInstance(InstancePath path, InstanceOp inst);
+};
+
+/// Necessary information to wire up a port with tapped data or memory location.
+struct PortWiring {
+  unsigned portNum;
+  ArrayRef<InstancePath> prefices;
+  SmallString<16> suffix;
+};
+
+} // namespace
+
+ArrayRef<InstancePath> InstancePaths::getAbsolutePaths(Operation *op) {
+  assert((isa<FModuleOp, FExtModuleOp>(op))); // extra parens makes parser smile
+
+  // If we have reached the circuit root, we're done.
+  if (op == instanceGraph.getTopLevelNode()->getModule()) {
+    static InstancePath empty{};
+    return empty; // array with single empty path
+  }
+
+  // Fast path: hit the cache.
+  auto cached = absolutePathsCache.find(op);
+  if (cached != absolutePathsCache.end())
+    return cached->second;
+
+  // For each instance, collect the instance paths to its parent and append the
+  // instance itself to each.
+  SmallVector<InstancePath, 8> extendedPaths;
+  for (auto inst : instanceGraph[op]->uses()) {
+    auto instPaths = getAbsolutePaths(inst->getParent()->getModule());
+    extendedPaths.reserve(instPaths.size());
+    for (auto path : instPaths) {
+      extendedPaths.push_back(appendInstance(path, inst->getInstance()));
+    }
+  }
+
+  // Move the list of paths into the bump allocator for later quick retrieval.
+  ArrayRef<InstancePath> pathList;
+  if (!extendedPaths.empty()) {
+    auto paths = allocator.Allocate<InstancePath>(extendedPaths.size());
+    std::copy(extendedPaths.begin(), extendedPaths.end(), paths);
+    pathList = ArrayRef<InstancePath>(paths, extendedPaths.size());
+  }
+
+  absolutePathsCache.insert({op, pathList});
+  return pathList;
+}
+
+InstancePath InstancePaths::appendInstance(InstancePath path, InstanceOp inst) {
+  size_t n = path.size() + 1;
+  auto newPath = allocator.Allocate<InstanceOp>(n);
+  std::copy(path.begin(), path.end(), newPath);
+  newPath[path.size()] = inst;
+  return InstancePath(newPath, n);
+}
+
+/// Return a version of `path` that skips all front instances it has in common
+/// with `other`.
+static InstancePath stripCommonPrefix(InstancePath path, InstancePath other) {
+  while (!path.empty() && !other.empty() && path.front() == other.front()) {
+    path = path.drop_front();
+    other = other.drop_front();
+  }
+  return path;
+}
```

Comment:
That sounds pretty interesting, wasn't aware of this approach.

I'm not happy at all with the common prefix stripping, or rather trying to guess what module/instance the user actually intended with an annotation. Maybe we can make these target instances instead of modules, to more clearly communicate what the user is interested in seeing.